rules/ast/ast.go

```diff
@@ -29,7 +29,12 @@ import (
 	"github.com/prometheus/prometheus/storage/metric"
 )
 
-var stalenessDelta = flag.Duration("query.staleness-delta", 300*time.Second, "Staleness delta allowance during expression evaluations.")
+var (
+	stalenessDelta = flag.Duration("query.staleness-delta", 300*time.Second, "Staleness delta allowance during expression evaluations.")
+	queryTimeout = flag.Duration("query.timeout", 2*time.Minute, "Maximum time a query may take before being aborted.")
+)
+
+var queryTimeoutErr = fmt.Errorf("query timeout after %v", *queryTimeout)
```

Comment:
IMHO, we really want the user to be able to increase the query timeout. (Prometheus is not really designed to serve untrusted users.) It's quite reasonable that you want to run one super heavy query, taking the informed risk, without changing the flags of the server.

I guess we can solve the per-query timeout in a later PR. To have a timeout at all is already a huge win.